### PR TITLE
Russell - L242.Bld_Inc_Elas_scenarios

### PR DESCRIPTION
**Pull request by russellhz**
There is one thing that we might want to change: the script uses A42.demand.csv (which is a pretty bizarre file) but only takes its sector name of “buildings”. It doesn’t actually use the data (if you can even call it data). Maybe we should drop this file.

#291

### CODE REVIEW

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R [hunk 1]
```diff
@@ -34,55 +44,185 @@ module_socioeconomics_L242.Bld_Inc_Elas_scenarios_DISABLED <- function(command,
...
+      rename(population = value) %>%
+      left_join_error_no_match(L102.gdp_mil90usd_GCAM3_R_Y, by = c("GCAM_region_ID","year")) %>%
+      mutate(pcgdp_90thousUSD = gdp/population) %>%
+      select(-population, -gdp) %>%
+      filter(year %in% FUTURE_YEARS)
+
+    # Linearly interpolate income elasticity for each level of per-capita GDP,
+    # using the assumption data
+    L242.IncomeElasticity_bld_GCAM3 <- L242.pcgdp_GCAM3_R_Y %>%
+      mutate(income.elasticity = approx(x = A42.inc_elas$pcgdp_90thousUSD, y = A42.inc_elas$inc_elas,
```
Comment (bpbond) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: Why not use the existing `gcamdata::approx_fun` (still need to change that name!)?

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R [hunk 2]
```diff
@@ -34,55 +44,185 @@ module_socioeconomics_L242.Bld_Inc_Elas_scenarios_DISABLED <- function(command,
...
+                                        # data will be assigned the cloest data extreme
+                                        rule = 2)$y %>% round(3),
+             energy.final.demand = A42.demand$energy.final.demand) %>%
+      select(scenario, region, energy.final.demand, year, income.elasticity) %>%
+      arrange(year)
+
+    # Split by scenario and remove scenario column from each tibble
+
+    L242.pcgdp_thous90USD_Scen_R_Y <- L242.pcgdp_thous90USD_Scen_R_Y %>%
+      split(.$scenario) %>%
```
Comment (bpbond) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: Huh. I totally didn't understand what was happening here. Now do I think.

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R [hunk 3]
```diff
@@ -34,55 +44,185 @@ module_socioeconomics_L242.Bld_Inc_Elas_scenarios_DISABLED <- function(command,
...
+      add_comments("Uses previously calculated per-capita GDP assumptions for all SSP scenarios") %>%
+      add_comments("Building income elasticity for each GCAM region generated by linear interpolation of assumption data") %>%
+      add_legacy_name("L242.IncomeElasticity_bld_gSSP1") %>%
+      add_precursors("common/GCAM_region_names", "energy/A42.demand", "socioeconomics/A42.inc_elas",
+                     "temp-data-inject/L102.pcgdp_thous90USD_Scen_R_Y") ->
+      L242.IncomeElasticity_bld_gSSP1
+
+    L242.pcgdp_thous90USD_Scen_R_Y[["gSSP2"]] %>%
+      add_title("Building Income Elasticity: gSSP2") %>%
+      add_units("Unitless (% change in service demand / % change in income)") %>%
```
Comment (bpbond) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: Given how much repetition there is, does it make sense to put some of this repeated stuff in lines 94-98 above?

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R (hunk 1 above)
Comment (russellhz) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: I may be wrong in my understanding of approx_fun, but I don't think that would work here. This is taking the gdp and income elasticity from another file and using that relationship to estimate income elasticity for different gdp data. Basically creating a linear model to generate predictions. That doesn't look to me what approx_fun is capable of handling.

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R (hunk 3 above)
Comment (russellhz) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: Hadn't thought of that. I will add that in.

**Comment by russellhz** [comment]
@bpbond This is one is under the socioeconomics team, not my normal energy team. Does that mean we should have different reviewers?

**Comment by bpbond** [comment]
Sorry, yes. Please change. Thanks!

On May 2, 2017, at 8:19 AM, russellhz <notifications@github.com<mailto:notifications@github.com>> wrote:


@bpbond<https://github.com/bpbond> This is one is under the socioeconomics team, not my normal energy team. Does that mean we should have different reviewers?

File: inst/extdata/socioeconomics/A42.inc_elas.csv [hunk 4]
```diff
@@ -0,0 +1 @@
...
15,0.6
20,0.5
25,0.4
30,0.33
35,0.27
40,0.2
45,0.15
50,0.1
55,0.05
60,0
```
Comment (rplzzz) [inline inst/extdata/socioeconomics/A42.inc_elas.csv]: Could somebody convert the line endings on this before final merge?

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R [hunk 5]
```diff
@@ -1,88 +1,181 @@
...
+
+module_socioeconomics_L242.Bld_Inc_Elas_scenarios <- function(command, ...) {
   if(command == driver.DECLARE_INPUTS) {
     return(c(FILE = "common/GCAM_region_names",
-             FILE = "energy/A42.demand",
              FILE = "socioeconomics/A42.inc_elas",
-             "L101.Pop_thous_GCAM3_R_Y",
-             "L102.gdp_mil90usd_GCAM3_R_Y",
-             "L102.pcgdp_thous90USD_Scen_R_Y"))
+             FILE = "L101.Pop_thous_GCAM3_R_Y",
```
Comment (d3y419) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: tagging for GCAM 3 legacy. no immediate action needed.

**Comment by d3y419** [comment]
this is a rarely used piece of code that replicates MiniCAM structure. 
looks like the code does what it should. i approve this code. 
two potential broader issues:
1. per @rplzzz , we're getting rid of GCAM3 legacy. this one includes it. 
2. per @bpbond https://github.com/JGCRI/gcamdata/issues/321 , this one doesn't have the necessary level 2 IDstring header to create xml input.

File: R/zchunk_L242.Bld_Inc_Elas_scenarios.R [hunk 6]
```diff
@@ -1,88 +1,181 @@
...
+      add_title("Building Income Elasticity: SSP4") %>%
+      add_legacy_name("L242.IncomeElasticity_bld_SSP4") ->
+      L242.IncomeElasticity_bld_SSP4
+
+    L242.pcgdp_thous90USD_Scen_R_Y[["SSP5"]] %>%
+      add_title("Building Income Elasticity: SSP5") %>%
+      add_legacy_name("L242.IncomeElasticity_bld_SSP5") ->
+      L242.IncomeElasticity_bld_SSP5
+
+    return_data(L242.IncomeElasticity_bld_GCAM3,
```
Comment (rplzzz) [inline R/zchunk_L242.Bld_Inc_Elas_scenarios.R]: As Haewon pointed out, we are no longer building the GCAM3 scenarios, so this output should be removed, along with all of the code to build it.

File: inst/extdata/socioeconomics/A42.inc_elas.csv (hunk 4 above)
Comment (bpbond) [inline inst/extdata/socioeconomics/A42.inc_elas.csv]: Done.

**Comment by russellhz** [comment]
I removed GCAM3 and switched one of the inputs from temp-data-inject. Looks like maybe we still need to wait to figure out how to deal with level 2 code?

**Comment by bpbond** [comment]
@russellhz yeah that's probably true. I'll try to nail that down tomorrow.

**Comment by d3y419** [comment]
@rplzzz : approved?

**Comment by bpbond** [comment]
Note - Level 2 ID strings not an issue here (another chunk will handle XML generation; see #358 ). I will add required files to this PR once we approve and merge #358 .

**Pull request merged**